libminifi/include/utils/MinifiConcurrentQueue.h

```diff
@@ -21,14 +21,32 @@
 #include <deque>
 #include <mutex>
 #include <condition_variable>
+#include <utility>
 #include <stdexcept>
+#include <type_traits>
 
 namespace org {
 namespace apache {
 namespace nifi {
 namespace minifi {
 namespace utils {
 
+namespace detail {
+template<typename...>
+using void_t = void;
+
+template<typename /* FunType */, typename T, typename = void>
+struct TryMoveCall {
+    template<typename Fun>
+    static void call(Fun&& fun, T& elem) { std::forward<Fun>(fun)(elem); }
+};
+
+template<typename FunType, typename T>
+struct TryMoveCall<FunType, T, void_t<decltype(std::declval<FunType>()(std::declval<T>()))>> {
```

Comment:
Agreed.

Reply:
Typo in 4.: s/cosidered/considered/
2.: makes no assumptions about the parameter type of a `FunType` object. If it takes an lvalue, then passing 1. will be ill-formed.

Rewritten in my words, but significantly longer and more complex. Take what you like:
1. `std::declval` looks similar to this: `template<typename T> T&& declval();`. Not defined, therefore it's only usable in unevaluated context. No requirements regarding `T`, therefore makes it possible to create hypothetical objects without requiring e.g. a default constructor and a destructor, like the `T{}` expression does.
2. `std::declval<FunType>()` resolves to an object of type `FunType`. If `FunType` is an lvalue reference, then this will also result in an lvalue reference due to reference collapsing.
3. `std::declval<FunType>()(std::declval<T>())` resolves to an object of the result type of a call on a function object of type `FunType` with an rvalue argument of type `T`. It is ill-formed if the function object expect an lvalue reference.
    - Example: `FunType` is a pointer to a `bool(int)` and `T` is `int`. This expression will result in a `bool` object.
    - Example: `FunType` is a function object modeling `bool(int&)` and `T` is `int`. This expression will be ill-formed because it's illegal to bind an `int` rvalue to an `int&`.
4. `void_t<decltype(*3*)>` checks for the well-formedness of 3., then discards it. If 3. is ill-formed, then this specialization is ignored through SFINAE. If well-formed, then it's considered more specialized than the other and takes precedence.

Reply:
Will perform a capture-by-copy-paste on it, this is as good as a documentation can get :)

Reply:
Comment updated in the commit mentioning your name :)